### DISCUSSION TITLE
Generic regex managers break if using docker datasource with digest pinning enabled

### DISCUSSION THREAD

**Discussion by maxbrunet** [renovatebot/renovate#24586] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

When using a generic regex manager that does not support the `currentDigest` capture group while using the `docker:pinDigests` preset is used, it breaks entirely.

Example regex:

```regex
# renovate: datasource=(?<datasource>.+?) depName=(?<depName>.+?)(?: (?:packageName)=(?<packageName>.+?))?(?: versioning=(?<versioning>.+?))?(?: extractVersion=(?<extractVersion>.+?))?\\s[A-Z_]+?_VERSION=('|\")?(?<currentValue>.+?)('|\")?\\s
```

Example text to match:

```shell
# renovate: datasource=docker depName=docker.io/goreleaser/goreleaser-cross
GOLANG_CROSS_VERSION=v1.21.1
```

It is not possible to add `currentDigest` support to this type of manager, as they can be used with various datasources (e.g. `github-tags` where the digest is Git commit and the usage would be different).

Would it be possible to detect the absence vs the emptiness of the `currentDigest` capture group in the regex and ignore pinning in those cases?

This kind of regex managers used to work in this configuration, the change must have happened in the past few weeks. It now blocks updates for matching packages and the entire `renovate/pin-dependencies` branch for affected repositories.

Minimal reproduction: https://github.com/maxbrunet/renovate-regex-no-docker-digest

### Relevant debug logs

<details><summary>Logs</summary>

```jsonc
DEBUG: Digest is not updated (branch="renovate/pin-dependencies")
{
  "manager": "regex"
  "packageFile": "script.sh"
  "expectedValue": "sha256:ee3b56b3f42a95dc953ede75f794e73bb3723dbc6ec1661bdeb45723918ce301"
}

WARN: Error updating branch: update failure (branch="renovate/pin-dependencies")
```

</details>

https://developer.mend.io/github/maxbrunet/renovate-regex-no-docker-digest/-/job/f5e03d49-8a6e-4015-b70a-7bb3e1449de9

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction in the description above

**Comment by rarkins** [comment]
Are you saying that there's no way to achieve pinDigests in the regex manager?

My concern with blocking digest pinning in the regex manager altogether is that maybe some users have managed to get it to work in some cases and now we'd block that

**Reply by viceice** [reply]
it works, but needs manual config.

you need to set the auto replace template correctly

**Reply by maxbrunet** [reply]
> it works, but needs manual config.
> 
> you need to set the auto replace template correctly

This might not be the best UX, but could you elaborate on this, please? And how would it work for [the presets](https://docs.renovatebot.com/presets-regexManagers/)?

> My concern with blocking digest pinning in the regex manager altogether is that maybe some users have managed to get it to work in some cases and now we'd block that

I would not want to block it altogether, but find a way to do it when it makes sense only, maybe checking something like `matchString.includes('(?<currentDigest>')`?

**Reply by rarkins** [reply]
But isn't this pinDigests, so currentDigest doesn't exist?

**Reply by maxbrunet** [reply]
Yes, but if the regex does not support matching it, I do not think it makes sense to try to pin it.

Also sometimes the `docker` datasource is used in a non-docker context, because it is the only source for updates for a piece of software (e.g. `golang` before it got its own datasource, or even [python](https://github.com/renovatebot/renovate/blob/5a208cfd2d00aa251999c2d937d70a745a9bb307/lib/modules/manager/pyenv/extract.ts#L9) internally, thus likely to happen in regex managers too), in that case, the digest is simply not useful.

**Reply by rarkins** [reply]
Regex can support pinning it. It's just the one you wrote does not

**Reply by rarkins** [reply]
Recommendation: Enable digest pinning based on manager, not datasource. Or manager + datasource.

**Reply by maxbrunet** [reply]
> Regex can support pinning it. It's just the one you wrote does not

Yes, correct: when a regex is not written to support it, it should not try

> Recommendation: Enable digest pinning based on manager, not datasource. Or manager + datasource.

hmm sure, would that mean the `regex` manager would have it disabled by default?

The goal would be to have [all these presets](https://github.com/cohere-ai/technical-architecture/pull/25) working when `datasource=docker` and `docker:pinDigests` are used (whatever pinning when `currentDigest` is supported happens or not is to be decided)

**Reply by fgreinacher** [reply]
I don't fully understand why this stopped working, can you point me to the relevant PR? 

Edit: Could it be related to https://github.com/renovatebot/renovate/issues/24135?

**Reply by fgreinacher** [reply]
>> it works, but needs manual config.
>> you need to set the auto replace template correctly
>
> This might not be the best UX, but could you elaborate on this, please? And how would it work for [the presets](https://docs.renovatebot.com/presets-regexManagers/)?

Same question for me. Based on the discussion so far I don't know what's missing in the config. Would be cool to get a bit more input @viceice 🙇

**Reply by fgreinacher** [reply]
We're currently working around this by disabling digest pinning per file:

```json
{
  "description": "Don't pin Docker digests in var files",
  "matchFileNames": ["group_vars/**"],
  "matchDatasources": ["docker"],
  "pinDigests": false
}
```

That works but does not scale well. Can we somehow fix this in the custom manager? As mentioned above this will also break many users of the [RegexManager presets](https://docs.renovatebot.com/presets-regexManagers/).

**Reply by rarkins** [reply]
Maybe I'm misunderstanding the problem. Please create a minimum reproduction of the _problem_ and describe also its impact. e.g. is the impact that you get a warning message and that's an annoyance, or does it _stop things working_?

**Reply by rarkins** [reply]
When I look at the reproduction above, I see this log:

```
DEBUG: packageFiles with updates (repository=renovate-reproductions/24586, baseBranch=main)
       "config": {
         "regex": [
           {
             "deps": [
               {
                 "depName": "docker.io/goreleaser/goreleaser-cross",
                 "currentValue": "v1.21.1",
                 "datasource": "docker",
                 "replaceString": "# renovate: datasource=docker depName=docker.io/goreleaser/goreleaser-cross\nGOLANG_CROSS_VERSION=v1.21.1\n",
                 "updates": [
                   {
                     "isPinDigest": true,
                     "updateType": "pinDigest",
                     "newValue": "v1.21.1",
                     "newDigest": "sha256:ee3b56b3f42a95dc953ede75f794e73bb3723dbc6ec1661bdeb45723918ce301",
                     "branchName": "renovate/pin-dependencies"
                   }
                 ],
                 "packageName": "docker.io/goreleaser/goreleaser-cross",
                 "versioning": "docker",
                 "warnings": [],
                 "sourceUrl": "https://github.com/goreleaser/goreleaser-cross",
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "v1.21.1",
                 "fixedVersion": "v1.21.1"
               }
             ],
             "matchStrings": [
               "# renovate: datasource=(?<datasource>.+?) depName=(?<depName>.+?)(?: (?:packageName)=(?<packageName>.+?))?(?: versioning=(?<versioning>.+?))?(?: extractVersion=(?<extractVersion>.+?))?\\s[A-Z_]+?_VERSION=('|\")?(?<currentValue>.+?)('|\")?\\s"
             ],
             "packageFile": "script.sh"
           }
         ]
       }
```

So:
- There are no updates, only an attempted pin digest
- pinDigest does not work
- A warning message is issued

Now if I roll back the Go version to 1.20.0, I get this PR: https://github.com/renovate-reproductions/24586/pull/1/files

How is this "breaks entirely". Isn't this "works correctly although there's a log warning you can get rid of if you update your config"?

**Reply by fgreinacher** [reply]
I created a small reproduction at https://github.com/fgreinacher/renovate-regex-docker-pin just using built-in presets.

Also here the issue occurs in the `renovate/pin-dependencies` branch: https://developer.mend.io/github/fgreinacher/renovate-regex-docker-pin/-/job/52f46a9d-22db-482e-8a5c-4f4df5165fe9.

It's understandable why this happens. But it's not intuitive how to get out of it unless you have a deep understanding of how Renovate works.

I'm not sure how to improve the experience here though.

**Reply by fgreinacher** [reply]
Could the Regex patterns be improved so that a PR with the digest can be created? Then people would at least recognize that Renovate wants to pin it. If that does not make sense in the specific file they'd know they need to disable digest pinning.

Or should we disable digest pinning in the presets to prevent this?

```diff
{
  "customManagers": [
    {
      "customType": "regex",
      "datasourceTemplate": "docker",
      "fileMatch": [
        "(^|/)Chart\\.yaml$"
      ],
      "matchStrings": [
        "#\\s*renovate: image=(?<depName>.*?)\\s+appVersion:\\s*[\"']?(?<currentValue>[\\w+\\.\\-]*)"
      ],
+     "pinDigests": false
    }
  ]
}
```

**Reply by rarkins** [reply]
@fgreinacher thanks for the reproduction. What do you think is main _problem_ demonstrated? There's an Error in the Dependency Dashboard, however I consider that to be cosmetic. Is there a PR being held back which shouldn't be?

**Reply by maxbrunet** [reply]
The PR for the entire `renovate/pin-dependencies` branch is blocked even if the error is for a single dependency. I have updated my reproduction with an npm package:

> ## Errored
> 
> These updates encountered an error and will be retried. Click on a checkbox below to force a retry now.
> 
>  - [ ] Pin dependencies (`babel-core`, `docker.io/goreleaser/goreleaser-cross`)

**Reply by fgreinacher** [reply]
@rarkins What @maxbrunet says above is the only real functional issue. But I think with a more actionable error message (or a PR that (wrongly) adds the digest) it would be easier to solve.

**Comment by rarkins** [comment, marked as answer]
Created bug issue #24942
